### PR TITLE
Add fmpz power fmpz and UInt 

### LINKED ISSUE #967: fmpz^fmpz
Is there a reason that raising an fmpz to an fmpz is not allowed, e.g. ZZ(2)^ZZ(2)?

### PR DESCRIPTION
**Pull request by albinahlback**
Resolves #967

### CODE REVIEW

**Review by tthsqe12 — COMMENTED** [review]
You have to check the return of fmpz_pow_fmpz and throw if it fails. Also, a @test_throws ... fmpz(2)^fmpz(2)^100 in the test code would be good.

**Comment by albinahlback** [comment]
> You have to check the return of fmpz_pow_fmpz and throw if it fails. Also, a @test_throws ... fmpz(2)^fmpz(2)^100 in the test code would be good.

Thank you for your insightful comment! I think I have discovered a bug, for my machine I get
```
julia> a = fmpz(2)
2

julia> a^(a^60)
ERROR: OutOfMemoryError()
Stacktrace:
 [1] ^(::fmpz, ::fmpz) at /home/albin/jlpkg/Nemo.jl/src/flint/fmpz.jl:662
 [2] top-level scope at REPL[12]:1

julia> a^(a^100)
0
```
There seems to be a threshold for when the `OutOfMemoryError()` is no longer generated. I'm trying to look at the FLINT-code atm.

**Comment by tthsqe12** [comment]
It is difficult to know how much memory the user has. The number 2^2^60 might fit in the 64 bit address space, but no one has this much memory. 2^2^100 definitely won't fit. There are more issues with the size limitations of the mpz_type...

**Comment by albinahlback** [comment]
> It is difficult to know how much memory the user has. The number 2^2^60 might fit in the 64 bit address space, but no one has this much memory. 2^2^100 definitely won't fit. There are more issues with the size limitations of the mpz_type...

Hmm, yeah, I've noticed something similair for `^(::fmpz, ::Int)`. I will simply do `a^(a^60)` instead.

**Comment by albinahlback** [comment]
I hope this is okay.

**Comment by thofma** [comment]
I have not tested it myself, but should we be worried that `a^(a^100) == 0`?

**Comment by albinahlback** [comment]
> I have not tested it myself, but should we be worried that `a^(a^100) == 0`?

I should've understood the FLINT-code better. Anyway, now it should throw an error.

It would be nice if the FLINT function `fmpz_pow_ui` would return an `Int`, instead of `void`. But that probably should be discussed under that repo instead of this one.

**Comment by wbhart** [comment]
The Flint interface follows the GMP interface for mpz_pow_ui. It's very unlikely we'll change this.

**Comment by thofma** [comment]
This is now throwing a SIGABRT on Windows, which is rather bad. I actually don't understand how this can be happening, since the GMP shipped with julia >= 1.3 is patched to prevent this (https://github.com/JuliaLang/julia/blob/c1c52930bb58605b60ae723c28f63e43b73944f6/deps/patches/gmp_alloc_overflow_func.patch). Do you have an idea @tthsqe12?

P.S.: Here is what julia does for BigInt^BigInt: https://github.com/JuliaLang/julia/blob/master/base/gmp.jl#L581

**Comment by tthsqe12** [comment]
Since you are doing all of the checks for easy cases in julia anyways, I see little point in using fmpz_pow_fmpz now. Better to just coerce the exponent to an Int and let the inexact error handle it (like hecke).

**Comment by tthsqe12** [comment]
I wouldnt test for out of memory error on an example that does technically fit in the address space. Blowing up the memory on a test set is not a good idea. Simply checking that 2^2^100 throws will be good enough since this cannot fit in any computer for the foreseeable future.

**Comment by albinahlback** [comment]
@tthsqe12, please feel free to remove the ccall to `fmpz_pow_fmpz`.

I removed the test that blows up the memory.

**Comment by albinahlback** [comment]
> Since you are doing all of the checks for easy cases in julia anyways, I see little point in using fmpz_pow_fmpz now. Better to just coerce the exponent to an Int and let the inexact error handle it (like hecke).

@tthsqe12 do you want me to do this instead?

**Comment by tthsqe12** [comment]
Yes, you can do this yourself. So, we don't need {fmpz|fmpq}_pow_fmpz. However, powering in finite fields, fmpz_mod and nmod should all accept fmpz exponents. Not sure if they do in nemo currently

**Comment by albinahlback** [comment]
Merged the method for `Int` and `fmpz` since both converts to `UInt`. Moreover I think that `InexactError` is more fitting than `OutOfMemoryError`, since we never really test the memory.

Please let me know if I can improve the methods in any way!

File: src/flint/fmpz.jl [hunk 1]
```diff
@@ -627,7 +627,23 @@ function ^(x::fmpz, y::Int)
...
    else
       z = fmpz()
-      ccall((:fmpz_pow_ui, libflint), Nothing, (Ref{fmpz}, Ref{fmpz}, UInt), z, x, UInt(y))
+      ccall((:fmpz_pow_ui, libflint), Nothing,
+            (Ref{fmpz}, Ref{fmpz}, UInt), z, x, UInt(y))
+      z
+   end
+end
+
+function ^(x::fmpz, y::UInt)
```
Comment (tthsqe12) [inline src/flint/fmpz.jl:636]: I see no reason why this cannot be combined into the union above. The compiler should know that the condition y<0 is always false.

File: test/flint/fmpz-test.jl [hunk 2]
```diff
@@ -343,36 +343,37 @@ end
...
+      @test isone(a^e) && isone(a^fmpz(e))
+      @test a^e !== a^fmpz(e) !== a
    end
 
    a = fmpz(-1)
    for e = [-3, -1, 1, 3, 5]
-      @test a^e == a
-      @test a^e !== a
+      @test a^e == a^fmpz(e) == a
+      @test a^e !== a^fmpz(e) == a
```
Comment (tthsqe12) [inline test/flint/fmpz-test.jl:373]: The purpose of the !== test is to make sure that ^ always returns a new object. Thus, this line should be the two
```
@test a^e !== a
@test a^fmpz(e) !== a
```

**Review by tthsqe12 — CHANGES_REQUESTED** [review]

File: test/flint/fmpz-test.jl (hunk 2 above)
Comment (albinahlback) [reply test/flint/fmpz-test.jl]: Thanks!

File: src/flint/fmpz.jl (hunk 1 above)
Comment (albinahlback) [reply src/flint/fmpz.jl]: Done, thanks!

**Review by tthsqe12 — APPROVED** [review]

**Comment by thofma** [comment]
Thanks. Will merge right before the next breaking release.

**Pull request merged**